transitions/multi_form_alerts.js

```diff
@@ -169,42 +182,73 @@ const validateConfig = () => {
   }
 };
 
-const getFilteredReports = (alert, latestReport) => {
+/**
+ * Returns { countedReportsIds, newReports, phones }.
+ */
+const getCountedReportsAndPhones = (alert, latestReport) => {
   return new Promise((resolve, reject) => {
     const script = vm.createScript(`(${alert.isReportCounted})(report, latestReport)`);
-    let reports = countReports([ latestReport ], latestReport, script);
     let skip = 0;
+    let countedReportsIds = [ latestReport._id ];
+    let newReports = [ latestReport ];
+    let phones = [ ...getPhonesOneReport(alert.recipients, latestReport) ];
     async.doWhilst(
       callback => {
-        const options = { skip: skip, limit: BATCH_SIZE };
-        fetchReports(latestReport.reported_date - 1, alert.timeWindowInDays, alert.forms, options)
-          .then(fetched => callback(null, fetched))
+        getCountedReportsAndPhonesBatch(script, latestReport, alert, skip)
+          .then(output => {
+            countedReportsIds = countedReportsIds.concat(output.countedReportsIds);
+            newReports = newReports.concat(output.newReports);
+            phones = phones.concat(output.phones);
+            callback(null, output.numFetched);
+          })
           .catch(callback);
       },
-      fetched => {
-        const countedReports = countReports(fetched, latestReport, script);
-        reports = reports.concat(countedReports);
+      numFetched => {
         skip += BATCH_SIZE;
-        return fetched.length === BATCH_SIZE;
+        return numFetched === BATCH_SIZE;
       },
       err => {
         if (err) {
           return reject(err);
         }
-        resolve(reports);
+        resolve({ countedReportsIds: countedReportsIds, newReports: newReports, phones: _.uniq(phones) });
       }
     );
   });
 };
 
+/**
+ * Returns Promise({ numFetched, countedReportsIds, newReports, phones }) for the db batch with skip value.
+ */
+const getCountedReportsAndPhonesBatch = (script, latestReport, alert, skip) => {
+  const options = { skip: skip, limit: BATCH_SIZE };
+  const output = {};
+  let countedReports;
+  return fetchReports(latestReport.reported_date - 1, alert.timeWindowInDays, alert.forms, options)
+    .then(fetched => {
+      output.numFetched = fetched.length;
+      countedReports = countReports(fetched, latestReport, script);
+      output.countedReportsIds = countedReports.map(report => report._id);
+      return countedReports;
+    })
+    .then(counted => {
+      output.newReports = counted.filter(report => !isReportAlreadyMessaged(report, alert.name));
+      return output.newReports;
+    })
+    .then((newReports) => {
+      output.phones = getPhones(alert.recipients, newReports);
+      return output;
+    });
```

Comment:
Sometimes I like breaking sync code up into different `.then` blocks but I think in this case it makes it more confusing because you have the `output` object being updated in each. It'd be clearer to do...

```js
  return fetchReports(latestReport.reported_date - 1, alert.timeWindowInDays, alert.forms, options)
    .then(fetched => {
      const countedReports = countReports(fetched, latestReport, script);
      const newReports = countedReports.filter(report => !isReportAlreadyMessaged(report, alert.name));
      return {
        numFetched: fetched.length,
        countedReportsIds: countedReports.map(report => report._id),
        newReports: newReports,
        phones: getPhones(alert.recipients, newReports)
      };
    });
```